File: backend/app/services/auth_service.py
"""Authentication, users, and invitation service."""

from __future__ import annotations

import hashlib
import hmac
import secrets
import smtplib
from dataclasses import dataclass
from datetime import UTC, datetime, timedelta
from email.message import EmailMessage
from typing import Any

from sqlalchemy import func

from app.core.config import settings
from app.core.db import get_db_session, row_to_dict
from app.models.user import Invite, User, UserSession


@dataclass
class AuthUser:
    id: int
    email: str
    full_name: str | None
    role: str
    is_active: bool
    created_at: str
    updated_at: str


def _utcnow() -> datetime:
    return datetime.now(tz=UTC)


def _iso(dt: datetime) -> str:
    return dt.isoformat()


def _hash_password(password: str, salt: bytes | None = None) -> str:
    if salt is None:
        salt = secrets.token_bytes(16)
    digest = hashlib.pbkdf2_hmac("sha256", password.encode(), salt, 200_000)
    return f"{salt.hex()}:{digest.hex()}"


def _verify_password(password: str, encoded: str) -> bool:
    try:
        salt_hex, digest_hex = encoded.split(":", 1)
    except ValueError:
        return False
    calc = _hash_password(password, bytes.fromhex(salt_hex)).split(":", 1)[1]
    return hmac.compare_digest(calc, digest_hex)


def _orm_to_user(obj: User) -> AuthUser:
    return AuthUser(
        id=int(obj.id),
        email=str(obj.email),
        full_name=obj.full_name,
        role=str(obj.role),
        is_active=bool(obj.is_active),
        created_at=str(obj.created_at),
        updated_at=str(obj.updated_at),
    )


class AuthService:
    @staticmethod
    def has_users() -> bool:
        with get_db_session() as session:
            count = session.query(func.count(User.id)).scalar()
            return bool(count and count > 0)

    @staticmethod
    def _users_count(session) -> int:
        return session.query(func.count(User.id)).scalar() or 0

    @staticmethod
    def create_user(email: str, password: str, full_name: str | None, role: str | None = None) -> AuthUser:
        normalized_email = email.strip().lower()
        now = _iso(_utcnow())
        with get_db_session() as session:
            existing = session.query(User).filter(User.email == normalized_email).first()
            if existing:
                raise ValueError("Email already registered")

            assigned_role = role or ("admin" if AuthService._users_count(session) == 0 else "viewer")
            password_hash = _hash_password(password)
            user = User(
                email=normalized_email,
                password_hash=password_hash,
                full_name=full_name,
                role=assigned_role,
                is_active=1,
                created_at=now,
                updated_at=now,
            )
            session.add(user)
            session.flush()
            session.refresh(user)
            return _orm_to_user(user)

    @staticmethod
    def authenticate(email: str, password: str) -> AuthUser | None:
        normalized_email = email.strip().lower()
        with get_db_session() as session:
            user = session.query(User).filter(User.email == normalized_email).first()
            if user is None or not bool(user.is_active):
                return None
            if not _verify_password(password, str(user.password_hash)):
                return None
            return _orm_to_user(user)

    @staticmethod
    def create_session(user_id: int) -> dict[str, str]:
        token = secrets.token_urlsafe(48)
        now = _utcnow()
        expires_at = now + timedelta(hours=settings.session_ttl_hours)
        with get_db_session() as session:
            sess = UserSession(
                user_id=user_id,
                token=token,
                expires_at=_iso(expires_at),
                created_at=_iso(now),
                last_seen_at=_iso(now),
            )
            session.add(sess)
        return {"token": token, "expires_at": _iso(expires_at)}

    @staticmethod
    def get_user_by_session_token(token: str) -> AuthUser | None:
        now = _iso(_utcnow())
        with get_db_session() as session:
            user = (
                session.query(User)
                .join(UserSession, UserSession.user_id == User.id)
                .filter(
                    UserSession.token == token,
                    UserSession.expires_at > now,
                    User.is_active == 1,
                )
                .first()
            )
            if user is None:
                return None
            session.query(UserSession).filter(UserSession.token == token).update(
                {"last_seen_at": now}
            )
            return _orm_to_user(user)

    @staticmethod
    def revoke_session(token: str) -> None:
        with get_db_session() as session:
            session.query(UserSession).filter(UserSession.token == token).delete()

    @staticmethod
    def update_profile(user_id: int, full_name: str | None) -> AuthUser:
        now = _iso(_utcnow())
        with get_db_session() as session:
            user = session.query(User).filter(User.id == user_id).first()
            if user is None:
                raise ValueError("User not found")
            user.full_name = full_name
            user.updated_at = now
            session.flush()
            return _orm_to_user(user)

    @staticmethod
    def list_users() -> list[dict[str, Any]]:
        with get_db_session() as session:
            users = session.query(User).order_by(User.email).all()
            return [
                {
                    "id": u.id,
                    "email": u.email,
                    "full_name": u.full_name,
                    "role": u.role,
                    "is_active": u.is_active,
                    "created_at": u.created_at,
                    "updated_at": u.updated_at,
                }
                for u in users
            ]

    @staticmethod
    def update_user_role(user_id: int, role: str) -> dict[str, Any]:
        if role not in {"admin", "viewer"}:
            raise ValueError("Invalid role")
        now = _iso(_utcnow())
        with get_db_session() as session:
            user = session.query(User).filter(User.id == user_id).first()
            if user is None:
                raise ValueError("User not found")
            user.role = role
            user.updated_at = now
            session.flush()
            return {
                "id": user.id,
                "email": user.email,
                "full_name": user.full_name,
                "role": user.role,
                "is_active": user.is_active,
                "created_at": user.created_at,
                "updated_at": user.updated_at,
            }

    @staticmethod
    def create_invite(
        email: str,
        role: str,
        invited_by_user_id: int,
        cluster_name: str | None = None,
        expires_hours: int | None = None,
    ) -> dict[str, Any]:
        if role not in {"admin", "viewer"}:
            raise ValueError("Invalid role")

        normalized_email = email.strip().lower()
        ttl = expires_hours if expires_hours and expires_hours > 0 else settings.invite_ttl_hours
        now = _utcnow()
        expires_at = now + timedelta(hours=ttl)
        token = secrets.token_urlsafe(40)

        with get_db_session() as session:
            invite = Invite(
                email=normalized_email,
                role=role,
                token=token,
                invited_by_user_id=invited_by_user_id,
                cluster_name=cluster_name,
                status="pending",
                expires_at=_iso(expires_at),
                created_at=_iso(now),
            )
            session.add(invite)
            session.flush()
            invite_dict = row_to_dict(invite)

        if invite_dict is None:
            raise RuntimeError("Failed to create invite")

        AuthService._send_invite_email(invite_dict)
        invite_dict["invite_url"] = f"{settings.app_base_url.rstrip('/')}/accept-invite?token={token}"
        return invite_dict

    @staticmethod
    def list_invites() -> list[dict[str, Any]]:
        now = _iso(_utcnow())
        with get_db_session() as session:
            session.query(Invite).filter(
                Invite.status == "pending",
                Invite.expires_at <= now,
            ).update({"status": "expired"})
            invites = session.query(Invite).order_by(Invite.created_at.desc()).all()
            return [row_to_dict(inv) for inv in invites]

    @staticmethod
    def accept_invite(token: str, password: str, full_name: str | None) -> dict[str, Any]:
        now = _utcnow()
        now_iso = _iso(now)

        with get_db_session() as session:
            invite = session.query(Invite).filter(Invite.token == token).first()
            if invite is None:
                raise ValueError("Invalid invite token")

            if invite.status != "pending":
                raise ValueError("Invite is no longer active")

            if str(invite.expires_at) <= now_iso:
                invite.status = "expired"
                raise ValueError("Invite has expired")

            email = str(invite.email)
            role = str(invite.role)

            existing_user = session.query(User).filter(User.email == email).first()
            if existing_user:
                existing_user.password_hash = _hash_password(password)
                if full_name is not None:
                    existing_user.full_name = full_name
                existing_user.role = role
                existing_user.updated_at = now_iso
                user_id = existing_user.id
            else:
                new_user = User(
                    email=email,
                    password_hash=_hash_password(password),
                    full_name=full_name,
                    role=role,
                    is_active=1,
                    created_at=now_iso,
                    updated_at=now_iso,
                )
                session.add(new_user)
                session.flush()
                user_id = new_user.id

            invite.status = "accepted"
            invite.accepted_at = now_iso

            user = session.query(User).filter(User.id == user_id).first()
            if user is None:
                raise RuntimeError("Failed to accept invite")
            auth_user = _orm_to_user(user)

        session_data = AuthService.create_session(auth_user.id)
        return {
            "user": AuthService.serialize_user(auth_user),
            "token": session_data["token"],
            "expires_at": session_data["expires_at"],
        }

    @staticmethod
    def serialize_user(user: AuthUser) -> dict[str, Any]:
        return {
            "id": user.id,
            "email": user.email,
            "full_name": user.full_name,
            "role": user.role,
            "is_active": user.is_active,
            "created_at": user.created_at,
            "updated_at": user.updated_at,
        }

    @staticmethod
    def _send_invite_email(invite: dict[str, Any]) -> None:
        invite_url = f"{settings.app_base_url.rstrip('/')}/accept-invite?token={invite['token']}"

        if not settings.smtp_host:
            return

        message = EmailMessage()
        message["From"] = settings.invite_from_email
        message["To"] = invite["email"]
        message["Subject"] = "You are invited to NATS JetStream Manager"
        message.set_content(
            "\n".join(
                [
                    "You have been invited to NATS JetStream Manager.",
                    f"Role: {invite['role']}",
                    f"Cluster: {invite.get('cluster_name') or 'default'}",
                    f"Accept invitation: {invite_url}",
                    f"Expires at: {invite['expires_at']}",
                ]
            )
        )

        with smtplib.SMTP(settings.smtp_host, settings.smtp_port, timeout=10) as smtp:
            if settings.smtp_use_tls:
                smtp.starttls()
            if settings.smtp_user:
                smtp.login(settings.smtp_user, settings.smtp_password)
            smtp.send_message(message)

File: backend/src/auth/auth.service.ts
import { Injectable, UnauthorizedException, ConflictException, NotFoundException, BadRequestException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, LessThan } from 'typeorm';
import { ConfigService } from '@nestjs/config';
import { randomBytes, pbkdf2Sync } from 'crypto';
import { User } from '../database/entities/user.entity';
import { UserSession } from '../database/entities/session.entity';
import { Invite } from '../database/entities/invite.entity';

const PBKDF2_ITERATIONS = 200_000;
const PBKDF2_KEY_LENGTH = 32;
const PBKDF2_DIGEST = 'sha256';

@Injectable()
export class AuthService {
  private readonly sessionTtlHours: number;
  private readonly inviteTtlHours: number;

  constructor(
    @InjectRepository(User)
    private readonly userRepository: Repository<User>,
    @InjectRepository(UserSession)
    private readonly sessionRepository: Repository<UserSession>,
    @InjectRepository(Invite)
    private readonly inviteRepository: Repository<Invite>,
    private readonly configService: ConfigService,
  ) {
    this.sessionTtlHours = this.configService.get<number>('SESSION_TTL_HOURS', 168);
    this.inviteTtlHours = this.configService.get<number>('INVITE_TTL_HOURS', 72);
  }

  hashPassword(password: string): { hash: string; salt: string } {
    const salt = randomBytes(16).toString('hex');
    const hash = pbkdf2Sync(
      password,
      salt,
      PBKDF2_ITERATIONS,
      PBKDF2_KEY_LENGTH,
      PBKDF2_DIGEST,
    ).toString('hex');
    return { hash, salt };
  }

  verifyPassword(password: string, storedHash: string): boolean {
    const [salt, hash] = storedHash.split(':');
    if (!salt || !hash) {
      return false;
    }
    const derivedHash = pbkdf2Sync(
      password,
      salt,
      PBKDF2_ITERATIONS,
      PBKDF2_KEY_LENGTH,
      PBKDF2_DIGEST,
    ).toString('hex');
    return derivedHash === hash;
  }

  async createUser(
    email: string,
    password: string,
    fullName?: string,
    role?: string,
  ): Promise<User> {
    const existing = await this.userRepository.findOne({ where: { email } });
    if (existing) {
      throw new ConflictException('A user with this email already exists');
    }

    // First user becomes admin
    if (!role) {
      const userCount = await this.userRepository.count();
      role = userCount === 0 ? 'admin' : 'viewer';
    }

    const { hash, salt } = this.hashPassword(password);
    const passwordHash = `${salt}:${hash}`;

    const user = this.userRepository.create({
      email,
      password_hash: passwordHash,
      full_name: fullName || undefined,
      role,
      is_active: true,
    });

    return await this.userRepository.save(user);
  }

  async authenticate(email: string, password: string): Promise<User> {
    const user = await this.userRepository.findOne({ where: { email } });
    if (!user) {
      throw new UnauthorizedException('Invalid email or password');
    }

    if (!user.is_active) {
      throw new UnauthorizedException('Account is disabled');
    }

    if (!this.verifyPassword(password, user.password_hash)) {
      throw new UnauthorizedException('Invalid email or password');
    }

    return user;
  }

  async createSession(userId: number): Promise<{ token: string; expiresAt: Date }> {
    const token = randomBytes(32).toString('hex');
    const expiresAt = new Date();
    expiresAt.setHours(expiresAt.getHours() + this.sessionTtlHours);

    const session = this.sessionRepository.create({
      token,
      user_id: userId,
      expires_at: expiresAt,
    });

    await this.sessionRepository.save(session);

    return { token, expiresAt };
  }

  async getUserBySessionToken(token: string): Promise<User | null> {
    const session = await this.sessionRepository.findOne({
      where: { token },
    });

    if (!session) {
      return null;
    }

    if (new Date() > new Date(session.expires_at)) {
      await this.sessionRepository.remove(session);
      return null;
    }

    const user = await this.userRepository.findOne({
      where: { id: session.user_id },
    });

    if (!user || !user.is_active) {
      return null;
    }

    // Update last_seen_at
    session.last_seen_at = new Date();
    await this.sessionRepository.save(session);

    return user;
  }

  async revokeSession(token: string): Promise<void> {
    await this.sessionRepository.delete({ token });
  }

  async updateProfile(userId: number, fullName?: string | null): Promise<User> {
    const user = await this.userRepository.findOne({ where: { id: userId } });
    if (!user) {
      throw new NotFoundException('User not found');
    }

    if (fullName !== undefined) {
      user.full_name = fullName ?? '';
    }

    user.updated_at = new Date();
    return await this.userRepository.save(user);
  }

  async listUsers(): Promise<User[]> {
    return this.userRepository.find({
      order: { created_at: 'ASC' },
    });
  }

  async updateUserRole(userId: number, role: string): Promise<User> {
    const user = await this.userRepository.findOne({ where: { id: userId } });
    if (!user) {
      throw new NotFoundException('User not found');
    }

    user.role = role;
    user.updated_at = new Date();
    return await this.userRepository.save(user);
  }

  async createInvite(
    email: string,
    role: string,
    invitedByUserId?: number,
    clusterName?: string,
    expiresHours?: number,
  ): Promise<Invite> {
    const existingUser = await this.userRepository.findOne({ where: { email } });
    if (existingUser) {
      throw new ConflictException('A user with this email already exists');
    }

    const token = randomBytes(32).toString('hex');
    const ttl = expiresHours ?? this.inviteTtlHours;
    const expiresAt = new Date();
    expiresAt.setHours(expiresAt.getHours() + ttl);

    const invite = this.inviteRepository.create({
      email,
      token,
      role,
      invited_by_user_id: invitedByUserId ?? undefined,
      cluster_name: clusterName ?? undefined,
      expires_at: expiresAt,
    });

    return await this.inviteRepository.save(invite);
  }

  async listInvites(): Promise<any[]> {
    const invites = await this.inviteRepository.find({
      order: { created_at: 'DESC' },
    });

    const now = new Date();
    return invites.map((invite) => {
      const isExpired = !invite.accepted_at && new Date(invite.expires_at) < now;
      return {
        ...invite,
        status: isExpired ? 'expired' : invite.status,
      };
    });
  }

  async acceptInvite(
    token: string,
    password: string,
    fullName?: string,
  ): Promise<{ user: User; session: { token: string; expiresAt: Date } }> {
    const invite = await this.inviteRepository.findOne({ where: { token } });
    if (!invite) {
      throw new NotFoundException('Invalid invite token');
    }

    if (invite.accepted_at) {
      throw new BadRequestException('Invite has already been accepted');
    }

    if (new Date() > new Date(invite.expires_at)) {
      throw new BadRequestException('Invite has expired');
    }

    const user = await this.createUser(invite.email, password, fullName, invite.role);

    invite.accepted_at = new Date();
    invite.status = 'accepted';
    await this.inviteRepository.save(invite);

    const session = await this.createSession(user.id);

    return { user, session };
  }
}
